### ISSUE TITLE
Strongbolt looks for Model to authorize against, even if authorization is skipped.

### ISSUE THREAD

**Issue by onkeld** [AnalyticsMediaGroup/strongbolt#28]
After setting `config.skip_controller_authorization_for "Devise::Sessions", "Devise::Registrations"` in the Strongbolt initializer, when trying to log in a user via Devise, I get the Error `Strongbolt::ModelNotFound in Devise::SessionsController#create `.

Looks like strongbolt still wants to know which model to use for authorization, even if it is not performing authorization on the controller in question.

**Comment by onkeld** [comment]
While trying to figure out some more about what happens, I stumbled across this:
First off, it doesn't matter if authorization on the devise controllers is skipped or not. Even with commented out `config.skip_controller_authorization_for "Devise::Sessions", "Devise::Registrations"`, I am able to access the login view and will get the error above when trying to submit.

I am also able to access the strongbolt views without being logged in as user and I can change or add user groups, roles and everything else without ever being asked for credentials.

That is after using the strongbolt:seed rake command to setup a full access user group with one existing user in it.
This existing user is however not logget in atm. So shouldn't strongbolt prevent access to everything without a logged in user?

**Comment by onkeld** [comment]
Next step:

If I request a password reset from devise and change my password, I am logged in as user afterwards. 
When then trying to edit my user from within, I get the error about missing model again.

**Comment by onkeld** [comment]
One further piece of information which might help solve the issue:
I have a PagesController I wrote myself.
On this I set `skip_controller_authorization` as I want to display a homepage to everybody, logged in or not. This seems to work fine enough.
Once I comment out this line, essentially asking Strongbolt to work on this controller again, I get my model error from before as this controller is model-less.
Telling the controller to authorize against the User model via `self.model_for_authorization = "User"` makes things work as expected again here.

Seeing this, I get the impression that skipping authorization from the initializer isn't doing anything at all. When I use the `config.skip_controller_authorization_for "Devise::Sessions", "Devise::Registrations"` line, I get the same behaviour as with my own controller when not skipping authorization but not telling strongbolt which model to authorize against.

And as I said before: I have full access to the strongbolt routes (e.g. http://localhost:3000/strongbolt/user_groups) without being logged in.
Kind of misses the point if I'm able to modify my roles/permissions without cheking there...
I'll open a new issue for that as well.

**Comment by tobithiel** [comment]
There is a lot of unrelated stuff to the actual issue here.
1. Strongbolt doesn't handle authentication but just handles authorization. So for any authentication related question please consult Devise or whatever you are using.
2. You should restart all your rails processes when you edit an initializer, if you're not already doing that. If you're using a Rails preloader like spring or others, look into their docs about what needs to be done.

**Comment by onkeld** [comment]
Thanks for the tips there.
Authentication with devise seems to work fine since I'm redirected to the login page on my own controllers when trying to access restricted methods.
The problem starts at that point, when trying to login and being prevented to do so by strongbolt.

After a full restart of the system I still get the model not found error on the devise controller when trying to log in.
I kind of understand where the error is coming from originally, as devise by design doesn't follow the rails name pattern - the model in my case being User and the controller being Devise::Sessions. So in my understanding the controller would need `self.model_for_authorization = "User"` to work properly. 
But all that shouldn't matter if you tell strongbolt not to authorize for a controller at all since there is no point in knowing what to authorize against then...

**Comment by tobithiel** [comment]
Yup, my point were just for the other things you mentioned above. The missing model thing shouldn't happen when authorization is skipped.

**Comment by onkeld** [comment]
Simple workaround for now (not tested yet, will try this out soon):
Devise allows for Controller customization.
So instead of using the 'devise for' standard routes, why not generate a custom controller via the 'rails generate devise:controllers [scope]' command for the User model.
Then, the custom controller would have a model with corresponding name to authenticate against. Added Bonus: we wouldn't have to use the skip authorization via the initializer but could simply add the directive to the controller.

Will try how this works and report back.

**Comment by onkeld** [comment]
Getting somewhere - workaround seems to work.
When generating a custom devise controller as suggested above, I am able to sign up and log in my devise user just fine, however I'm still not able to log out afterwards.
I'm not asked for a model anymore, but instead, Strongbolt tells me 'Strongbolt::Unauthorized in Users::SessionsController#destroy update permission not granted to User:1 for resource User:1'
even with 'skip_controller_authorization' on said sessions controller.

**Comment by onkeld** [comment]
In above case, if I seed the database with 'rake strongbolt:seed' to generate a full access user group with my user inside, I can edit and log out.
This tells me 'skip_controller_authorization' does apparently do nothing on my controller since I still have to be permitted to take actions.

**Comment by onkeld** [comment]
Going step by step to figure out, where the error comes from - devise or strongbolt.
I reset the database.
I reran 'rails g strongbolt:install' to get a fresh out of the box strongbolt initializer.
I restarted my computer to make sure there's no ruby processes left running anywhere.
I started rails server.

Situation: 
Fresh strongbolt without any seed for user groups or roles in the database.
No skip_authorization in the initializer or any of the controllers.
There is no current_user as there is no user in the database, let alone loaded by rails.
Strongbolt is aware of the User model - it appears in the generated strongbolt initializer.
Result: 
I am able to sign up a user using devise.
Even though strongbolt should bolt the User model (the generator added User to the list of models in the initializer).

Question: Why am I able to create a new database entry to the User model without strongbolt complaining?
Answer: Because I don't check for authentication on the controller first - so there is no current user to authorize by strongbolt. In the logfile it notes 'No authorization checking because no current user'.
If Strongbolt is supposed to lock down everything right from the start, shouldn't it especially lock things down if there is no user set?
But this is besides the point.

Next I tried to log in with the just created user and I am getting the 'Strongbolt::ModelNotFound in Devise::SessionsController#create'.
This is with not specifying any skip authorization - not on the controller itself and not in the initializer.

Same thing happens, if I set 'config.skip_controller_authorization_for "Users::Sessions"' in the initializer.

Here it gets strange since I have a Users::Sessions controller (the customized Devise controller generated before), which is connected to the User model - so there IS a model for strongbolt to authorize against. Strongbolt claims it doesn't find the User model.

If however, I set 'skip_controller_authorization' on the Users controller itself instead of setting it in the Initializer, everything works like a breeze - I can login my user, logout my user and edit my user since it belongs to the full access user group.
If I create a second user without full access, this one can log in and out just fine, but has no access to the edit user screen since it is 'Strongbolt::Unauthorized in Devise::RegistrationsController#edit'.

So to me it looks like there's something wrong in the Initializer variant of skipping authorization.

For me, the problem is solved here as I will need to use my own controllers going forward anyways and with the current setup things seem to work.
If you need further Information to work out the model not found stuff, let me know!

**Comment by tobithiel** [comment]
The `Strongbolt::ModelNotFound` error is definitely from Strongbolt and has nothing to do with Devise.
The reason your `Users::Sessions controller` throws a `Strongbolt::ModelNotFound` error, is that your controller name is not `<ModelName>Controller` which is what Rails defaults to. If you want to do that you can use `self.model_for_authorization = ...` to tell Strongbolt the model name and it won't try to search for it.
My guess for your login works but logout doesn't work scenario, is that `current_user` doesn't exist when Strongbolt makes its checks on login, but when the logout happens `current_user` does exist.
